man/zram-generator.md

```diff
@@ -0,0 +1,57 @@
+zram-generator(8) -- Systemd unit generator for zram swap devices
+=================================================================
+
+## SYNOPSIS
+
+`/lib/systemd/system-generators/zram-generator` `TARGET_DIR` [*2RGET_DIR* *3RGET_DIR*]<br />
+`/lib/systemd/system-generators/zram-generator` --setup-device `DEVICE`<br />
+`/lib/systemd/system-generators/zram-generator` --reset-device `DEVICE`
+
+## DESCRIPTION
+
+`zram-generator` is a generator that creates zram devices and systemd units to configure, format, and enable swap on them.
+
+The generator will be invoked by systemd early at boot, the generator will then:
+
+  1. read configuration files from *{/etc,/lib}/systemd/zram-generator.conf[.d]* (see zram-generator.conf(5) for details);
+  2. if systemd-detect-virt(1) exists and returns non-*0*, exit, since continuing would mean changing the host system;
+  3. generate systemd.swap(5) units into `TARGET_DIR` for all configured devices and symlink them into *swap.target.wants/*;
+  4. if any devices are enabled, and the zram module isn't loaded (*/sys/class/zram-control* doesn't exist), invoke *modprobe(8) zram*;
+  5. if the requested devices aren't present, read */sys/class/zram-control/hot_add* until the highest device's number is read,
+
+If the `ZRAM_GENERATOR_ROOT` environment variable is present, the return value, if any, of systemd-detect-virt(1) is ignored, and steps `4` and `5` are skipped;
+for the ramifications of `ZRAM_GENERATOR_ROOT` on config handling, see zram-generator.conf(5).
+
+
+The generated *dev-zramN.swap* units depend on `zram-create@zramN.service`, which on start invokes the generator with *--setup-device zramN*, which will then:
+
+  1. read configuration files from *{/etc,/lib}/systemd/zram-generator.conf[.d]* (see zram-generator.conf(5) for details);
+  2. set the desired compression algorithm in */sys/block/zramN/comp_algorithm*, if any;
+     if the write fails with `EINVAL` the current kernel doesn't understand the specified algorithm; a warning is issued, but execution continues as normal,
+     if the write fails for any other reason, setup is halted;
+  3. set the desired blockdev size in */sys/block/zramN/disksize*, if the write fails, setup is halted;
+  4. run *systemd-makefs(8) swap /dev/zramN*
+
+And on stop with *--reset-device zramN*, which will write a positive integer to */sys/block/zramN/reset*, allowing the device to be reused.
```

Comment:
I'd remove this paragraph and the previous one. Just describe the effect of the unit: the zram device is set up according to configuration, formatted as a swap device, and enabled. When the unit is stopped, the zram device is reset to free memory.

Reply:
I've simplified this sexion, but left the broad-strokes blow-by-blow, since it highlights (a) the unrecognised compression behaviour and (b) the potential race between generation and setup (e.g. if swap.target is started manually later).